### ISSUE TITLE
Model: Ensure generated IDs are globally unique

### ISSUE THREAD

**Issue by sebbert** [fusetools/fuselibs-public#512]

**Comment by kusma** [comment]
Are you sure they need to be *globally* unique?

**Comment by sebbert** [comment]
No, they need to be unique across model graphs. This ticket is for myself, so I didn't care too much about wording the problem in excruciating detail.

**Comment by sebbert** [comment]
Some more context:

The IDs are currently generated by incrementing the variable declared on this line: https://github.com/sebbert/fuselibs-public/blob/feature-Models/Source/Fuse.Reactive.JavaScript/FuseJS/Model.js#L22

If changing this to a string of some sort (a GUID for instance), TreeObservable also currently expects an ID of type `long`, so this needs to be changed as well: https://github.com/sebbert/fuselibs-public/blob/feature-Models/Source/Fuse.Reactive.JavaScript/TreeObservable/TreeObservable.uno#L80

**Comment by kristianhasselknippe** [comment]
After some discussion with sebbert, this doesn't seem like a pressing issue afterall:

Si at du har en app-modell med feltet `someObject`.
Model.js vil gi alle objektene en ID:

```class AppModel {
    constructor() { // AppModel-instansen vil få ID == 0
        this.someObject = { // someObject vil få ID == 1
            foo: 1337
        }
    }
}
```


Problemet er at denne IDen bare er unik innenfor den ene `Model=`-attributten.
Det jeg fryktet var at disse IDene vil kollidere:

Si at du så har en komponentmodell som tar inn `someObject` fra app-modellen som en ux:Property

```class MyComponent {
    constructor() { // MyComponent-instansen vil få ID == 0
        this.someOtherObject = {}; // someOtherObject vil få ID == 1
        this.SomeObject = null;
    }
}
```


`this.SomeObject` i kodesnutten over kommer fra et UX-property `SomeObject`:

```<Panel ux:Class="MyComponent" Model="MyComponent">
    <object ux:Property="SomeObject" />
</Panel>
```


Dette UX-propertyet blir matet med `someObject` fra app-modellen

```<App Model="AppModel">
    <MyComponent SomeObject="{someObject}">
</App>
```



Som du ser i de to modell-klassene over, vil både `someObject` fra app-modellen og `someOtherObject` fra komponent-modellen begge få samme ID (1). Problemet er at TreeObservable (og dermed Models) per i dag anser to objekter med samme ID som samme objekt (selv om de er to forskjellige JS-objekter). Ergo var jeg redd for at `MyComponent` vil anta at `someOtherObject === SomeObject`

MEN!

Dette viser seg å ikke være et problem allikevel.

Når Model.js gir en ID til et objekt, vil den ikke faktisk endre på objektet ditt. Den lager istedetfor en kopi av objektet med IDen og noe annen metadata, _samt en referanse til det opprinnelige objektet_.

Når objektet marshalles til Uno og tilbake, sitter vi igjen med det opprinnelige objektet, før det fikk en ID og annen metadata fra appmodellen. Dette vil si at komponentmodellen regner det som et nytt objekt, og lager sin egen ID for det.

Dette har i seg selv en del andre problemer though, f.eks. at vi kan endre på objektet i komponentmodellen uten at app-modellen plukker det opp og vise versa.

**Comment by kusma** [comment]
@kristianhasselknippe: Not everyone in the company speaks norwegian, would you care to translate?

**Comment by kristianhasselknippe** [comment]
Translation (an attempt at least):

Say you have an app-model with the field `someObject`.
Model.js would give all object an ID:

```
	constructor() { // AppModel-instansen vil få ID == 0
		this.someObject = { // someObject vil få ID == 1
			foo: 1337
        }
    }
}
```
The problem is that this ID is only unique within this one Model= attribute.
We feared that these IDs would collide between models.

Say you have a component model that take an object `someObject` from the app model as a ux:Property

```
    constructor() { // MyComponent-instansen vil få ID == 0
        this.someOtherObject = {}; // someOtherObject vil få ID == 1
        this.SomeObject = null;
    }
}
```

`this.SomeObject` in the above piece of code comes from the outside as a UX-property `SomeObject`.

```
    <object ux:Property="SomeObject" />
</Panel>
```

This UX-property will be fed with `someObject` from the app model.

```
    <MyComponent SomeObject="{someObject}">
</App>
```

As you can see in the two modell-classes above, `someObject` from the app model and `someOtherObject` from the component model will both have the same ID (1). The problem is that `TreeObservable` (and therefore Models) today views two objects with the same ID as the same object (even though they might be different JS objects). We were therefore afraig that `MyComponent` would assume that `someOtherObject === SomeObject`.

BUT!

This proved to not be a problem afterall.

When Model.js gives an ID to an object, it will not actually change your object. It will instead create a copy of the object with the ID and some other metadata, as well as a referance to the original object.

When the object is marshalled to Uno and back, what we are left with is the original object, before it got another ID and other metadata from the app model. This means that the componentmodel views it as a new object, and creates its own ID for it.

This does come with some other potential problems though, one example being that we can change the object in the component model without the app model picking up the change.
